Firestore support (#1657)
spring-cloud/spring-cloud-gcp: docs/src/main/asciidoc/firestore.adoc

== Cloud Firestore

> Spring Cloud GCP provides convenience starter which automatically configures authentication settings and client objects needed to begin using the https://cloud.google.com/firestore/[Google Cloud Firestore].

Reviewer: ```suggestion
Spring Cloud GCP provides a convenience starter which automatically configures authentication settings and client objects needed to begin using https://cloud.google.com/firestore/[Google Cloud Firestore].
```

> Spring Cloud GCP provides convenience starter which automatically configures authentication settings and client objects needed to begin using the https://cloud.google.com/firestore/[Google Cloud Firestore].

Reviewer: I think you need specify somewhere that this is for Firestore in **Native mode**.

See https://cloud.google.com/firestore/docs/[documentation] to learn more about Cloud Firestore.

> === Dependency Setup

Reviewer: I don't see the use of a separate section for "Dependency Setup" in other docs. We usually just make it part of the introductory section.
```suggestion
```

To begin using this library, add the `spring-cloud-gcp-starter-data-firestore` artifact to your project.

Maven coordinates, using <<getting-started.adoc#_bill_of_materials, Spring Cloud GCP BOM>>:

[source,xml]
----
<dependency>
    <groupId>org.springframework.cloud</groupId>

>     <artifactId>spring-cloud-gcp-starter-data-firestore</artifactId>

Reviewer: Why "data-firestore"?
I would just go with "spring-cloud-gcp-starter-firestore".

Author: It's for symmetry with spring-cloud-gcp-starter-data-datastore and spring-cloud-gcp-starter-data-spanner

</dependency>
----

Gradle coordinates:

[source]
----
dependencies {
  compile group: 'org.springframework.cloud', name: 'spring-cloud-gcp-starter-data-firestore'
}
----

=== Using Cloud Firestore

> In order to inject `Firestore` into your class, you can use `@Autowired` annotation.

Reviewer: ```suggestion
The starter automatically configures and registers a `Firestore` bean in the Spring application context. To start using it, simply use the `@Autowired` annotation.
```

[source,java]
----
@Autowired
Firestore firestore;

private void writeDocumentFromObject() throws ExecutionException, InterruptedException {
    // Add document data with id "joe" to collection "users" using a custom User class
    User data = new User("Joe",
            Arrays.asList(
                new Phone(12345, PhoneType.CELL),
                new Phone(54321, PhoneType.WORK)));

>     this.firestore.document("users/joe").set(data).get();

Reviewer: Can you split `get()` into a separate statement, and use its result somehow? Otherwise it's a bit confusing.

</dependency>
----

Gradle coordinates:

[source]
----
dependencies {
  compile group: 'org.springframework.cloud', name: 'spring-cloud-gcp-starter-firestore'
}
----

=== Sample

A https://github.com/spring-cloud-gcp/spring-cloud-gcp-samples/spring-cloud-gcp-firestore-sample[sample application] is available.

=== Using Cloud Firestore
The starter automatically configures and registers a `Firestore` bean in the Spring application context. To start using it, simply use the `@Autowired` annotation.

[source,java]
----
@Autowired
Firestore firestore;

private void writeDocumentFromObject() throws ExecutionException, InterruptedException {
    // Add document data with id "joe" to collection "users" using a custom User class
    User data = new User("Joe",
            Arrays.asList(
                new Phone(12345, PhoneType.CELL),
                new Phone(54321, PhoneType.WORK)));

    // asynchronously write data
    ApiFuture<WriteResult> result = docRef.set(data);

    // result.get() blocks on response
    LOGGER.info"Update time: " + result.get().getUpdateTime());
}

private void readDocumentToObject() throws ExecutionException, InterruptedException {

>     ApiFuture<DocumentSnapshot> documentSnapshotApiFuture = this.firestore.document("users/joe").get();

Author: `.get()` should be there, that is correct
Good point, I'll replace this with the executable code.
